src/base/static/libs/maps/mapboxgl-provider.js

```diff
@@ -549,35 +558,43 @@ export default (container, options) => {
 
     map.addSource(id, sourcesCache[id]);
 
-    rules = rules.map(rule => ({
-      baseLayerId: id,
-      source: id,
-      ...rule,
-    }));
-    focus_rules = focus_rules.map(rule => ({
-      baseLayerId: `${id}${FOCUSED_LAYER_IDENTIFIER}`,
-      source: FOCUSED_SOURCE_IDENTIFIER,
-      ...rule,
-    }));
-
-    // NOTE: We create a lot of layers here, which could be a performance
-    // bottleneck.
-    // See: https://github.com/mapseed/platform/issues/961
+    rules = rules
+      .map(rule => ({
+        baseLayerId: id,
+        source: id,
+        ...rule,
+      }))
+      .concat(
+        focus_rules.map(rule => ({
+          baseLayerId: `${id}${FOCUSED_LAYER_IDENTIFIER}`,
+          source: FOCUSED_SOURCE_IDENTIFIER,
+          ...rule,
+        })),
+      );
+
+    // TODO: Think about further optimizations for reducing/combining the
+    // number of provider layers created here.
     layersCache[id] = rules
-      .map(configRuleToSymbolLayer)
-      .concat(rules.map(configRuleToLineLayer))
+      .map(
+        feature_types.includes("Point") ? configRuleToSymbolLayer : () => null,
+      )
       .concat(
-        rules
-          .map(configRuleToFillLayer)
-          .reduce((flat, toFlatten) => flat.concat(toFlatten), []),
+        rules.map(
+          feature_types.includes("LineString")
+            ? configRuleToLineLayer
+            : () => null,
+        ),
       )
-      .concat(focus_rules.map(configRuleToSymbolLayer))
-      .concat(focus_rules.map(configRuleToLineLayer))
       .concat(
-        focus_rules
-          .map(configRuleToFillLayer)
+        rules
+          .map(
+            feature_types.includes("Polygon")
+              ? configRuleToFillLayer
+              : () => null,
+          )
           .reduce((flat, toFlatten) => flat.concat(toFlatten), []),
-      );
+      )
+      .filter(rule => !!rule);
```

Comment:
I'm not sure about the pattern of returning `null`s for unused `feature_types` and then filtering them out here. It works but seems clunky...

Reply:
Yeah, we should clean this up. See my suggestion above.